src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
 
         public Task Invoke(HttpContext httpContext)
         {
-            var feature = new EndpointSelectorContext();
+            var feature = new EndpointSelectorContext(httpContext);
+
+            // Store the previous endpoint before matching starts, this lets us restore whatever endpoint was
+            // set at the end of matching if nothing matched
+            var previousEndpoint = feature.Endpoint;
+
+            // We need to set the endpoint to null here so that we can observe any mutations
+            // made by the matcher
+            feature.Endpoint = null;
```

Comment:
Is there much perf impact with EndpointSelectorContext.Endpoint fetching the IEndpointFeature from HttpContext each time it gets and sets the endpoint?

What do you think of EndpointSelectorContext not wrapping HttpContext, and instead getting and setting the endpoint on the struct? The endpoint would then be set on HttpContext at the end of routing. Although would that require the struct be passed around by ref?

Reply:
The struct cannot be passed by ref as all of the methods are async, I don’t think it’ll have any real performance impact and if we wanted to solve this, I’d do it by hoisting the Endpoint to an HttpContext property

Reply:
This needs to preserve route values as well.

Reply:
Does it need to preserve route values if we short circuit here?